src/lib/storage/reference_segment.cpp

```diff
@@ -1,43 +1,45 @@
 #include "reference_segment.hpp"
 
 #include "storage/table.hpp"
+#include "types.hpp"
 #include "utils/assert.hpp"
 
 namespace opossum {
 
 ReferenceSegment::ReferenceSegment(const std::shared_ptr<const Table>& referenced_table,
-                                   const ColumnID referenced_column_id, const std::shared_ptr<const PosList>& pos) {
-  // Implementation goes here
-}
+                                   const ColumnID referenced_column_id, const std::shared_ptr<const PosList>& pos)
+    : _referenced_table{referenced_table}, _referenced_column_id{referenced_column_id}, _pos_list{pos} {}
 
 AllTypeVariant ReferenceSegment::operator[](const ChunkOffset chunk_offset) const {
-  // Implementation goes here
-  Fail("Implementation is missing.");
+  return get_by_row_id(_pos_list->at(chunk_offset));
+}
+
+AllTypeVariant ReferenceSegment::get_by_row_id(RowID row_id) const {
+  if (row_id.is_null()) {
+    return NULL_VALUE;
+  }
+  const auto chunk = _referenced_table->get_chunk(row_id.chunk_id);
+  return (*chunk->get_segment(_referenced_column_id))[row_id.chunk_offset];
 }
 
 ChunkOffset ReferenceSegment::size() const {
-  // Implementation goes here
-  Fail("Implementation is missing.");
+  return _pos_list->size();
 }
 
 const std::shared_ptr<const PosList>& ReferenceSegment::pos_list() const {
-  // Implementation goes here
-  Fail("Implementation is missing.");
+  return _pos_list;
 }
 
 const std::shared_ptr<const Table>& ReferenceSegment::referenced_table() const {
-  // Implementation goes here
-  Fail("Implementation is missing.");
+  return _referenced_table;
 }
 
 ColumnID ReferenceSegment::referenced_column_id() const {
-  // Implementation goes here
-  Fail("Implementation is missing.");
+  return _referenced_column_id;
 }
 
 size_t ReferenceSegment::estimate_memory_usage() const {
-  // Implementation goes here
-  Fail("Implementation is missing.");
+  return size() * sizeof(RowID);
```

Comment:
`PosList` is shared between `ReferenceSegment`s, beware of counting its memory twice. You also do not count `this` into the size.

Reply:
We are counting the entries of the one `PosList` just once, or am I not seeing this correctly? As for adding the size of `this`: We just stuck to what the tests for `ValueSegment` demanded back in sprint 1 which was just to account for the contained tuples. Either way seems reasonable to me